seq2struct/utils/parallelizer.py

```diff
@@ -0,0 +1,156 @@
+
+
+from abc import ABC, abstractmethod
+
+import time
+import queue
+import multiprocessing
+import tqdm
+
+class Parallelizer(ABC):
+    """
+    A parallelizer is a general purpose task used to handle the situation of
+        executing a function f(x, y) on a series of inputs x_i, y_i, where
+        x is often quite large and thus difficult to transport to a given
+        process, but there are typically fewer distinct values of x, allowing
+        for each process to use at most one value of x.
+
+    If it takes a constant amount of time `t_x` to ship x to a process and a constant
+        amount of time `t_y` to ship y to a process then execute f(x, y) and return the
+        result, then if we have `n` ys for a given x, then using `k` processes
+        takes time `k * t_x + n/k * t_y`, which is minimized when `t_x = n * t_y / k^2`,
```

Comment:
With how `multiprocessing` works, `t_x` can be divided into two parts: pickling then unpickling. The pickling can be done once, but the unpickling has to be done once per worker (and can be done concurrently between workers, in theory). That would change the model a little bit.
(this is more for intellectual curiosity, I don't think we need to get into implementing that right now)

Reply:
That's true, didn't think about that. I was including the execution time for f(x) on the other process, which I've removed